src/rust/src/backend/poly1305.rs

```diff
@@ -35,33 +76,60 @@ impl Poly1305 {
             ));
         }
 
-        #[cfg(all(not(CRYPTOGRAPHY_IS_LIBRESSL), not(CRYPTOGRAPHY_IS_BORINGSSL)))]
-        {
-            if cryptography_openssl::fips::is_enabled() {
-                return Err(CryptographyError::from(
-                    exceptions::UnsupportedAlgorithm::new_err((
-                        "poly1305 is not supported by this version of OpenSSL.",
-                        exceptions::Reasons::UNSUPPORTED_MAC,
-                    )),
-                ));
-            }
-
-            let pkey = openssl::pkey::PKey::private_key_from_raw_bytes(
-                key.as_bytes(),
-                openssl::pkey::Id::POLY1305,
-            )
-            .map_err(|_| {
-                pyo3::exceptions::PyValueError::new_err("A poly1305 key is 32 bytes long")
-            })?;
-
-            Ok(Poly1305 {
-                signer: Some(
-                    openssl::sign::Signer::new_without_digest(&pkey).map_err(|_| {
-                        pyo3::exceptions::PyValueError::new_err("A poly1305 key is 32 bytes long")
-                    })?,
-                ),
-            })
-        }
+        let pkey = openssl::pkey::PKey::private_key_from_raw_bytes(
+            key.as_bytes(),
+            openssl::pkey::Id::POLY1305,
+        )
+        .map_err(|_| pyo3::exceptions::PyValueError::new_err("A poly1305 key is 32 bytes long"))?;
+
+        Ok(Poly1305Open {
+            signer: Some(
+                openssl::sign::Signer::new_without_digest(&pkey).map_err(|_| {
+                    pyo3::exceptions::PyValueError::new_err("A poly1305 key is 32 bytes long")
+                })?,
+            ),
+        })
+    }
+    fn update(&mut self, data: CffiBuf<'_>) -> CryptographyResult<()> {
+        let buf = data.as_bytes();
+        self.get_mut_signer()?.update(buf)?;
+        Ok(())
+    }
+    fn finalize<'p>(
+        &mut self,
+        py: pyo3::Python<'p>,
+    ) -> CryptographyResult<&'p pyo3::types::PyBytes> {
+        let signer = self.get_mut_signer()?;
+        let result = pyo3::types::PyBytes::new_with(py, signer.len()?, |b| {
+            let n = signer.sign(b).unwrap();
+            assert_eq!(n, b.len());
+            Ok(())
+        })?;
+        self.signer = None;
+        Ok(result)
+    }
+}
+
+#[pyo3::prelude::pyclass(module = "cryptography.hazmat.bindings._rust.openssl.poly1305")]
+struct Poly1305 {
+    #[cfg(any(CRYPTOGRAPHY_IS_BORINGSSL, CRYPTOGRAPHY_IS_LIBRESSL))]
+    backend: Poly1305Boring,
+    #[cfg(not(any(CRYPTOGRAPHY_IS_LIBRESSL, CRYPTOGRAPHY_IS_BORINGSSL)))]
+    backend: Poly1305Open,
```

Comment:
I think you want to move the optionality up to this level, to reduce the duplication.